Add a mechanism to filter IPC messages by feature flags

rniwa opened the pull request:
#### f8b48e173616feeff41c1015c5d9406dac9c2f83
<pre>
Add a mechanism to filter IPC messages by feature flags
<a href="https://bugs.webkit.org/show_bug.cgi?id=276415">https://bugs.webkit.org/show_bug.cgi?id=276415</a>

Reviewed by Chris Dumez.

This PR adds a filtering mechanism for IPC messages based on which features are enabled at runtime.

To do this, this PR adds the code to generate SharedPreferencesForWebProcess.h/cpp/serialization.in,
which consists of a simple struct with a bunch of booleans indicating which preference is enabled
in a given WebContent process. UnifiedWebPreferences.yaml is now annotated with
&quot;sharedPreferenceForWebProcess&quot; flag per preference, and messages.in files now support &quot;EnabledBy=X&quot;
attribute which filters IPC messages by the preferences.

The set of preferences for a given WebContent process is synchronized at the time of constructing
GPUConnectionToWebProcess as a part of GPUProcessConnectionParameters, and updated via
GPUProcessProxy::filteringDidChangeForWebProcess. Because multiple WebPageProxy can share a single
WebContent process, the set is the union of all preferences enabled across those pages.

When a preference is updated, WebPageProxy::preferencesDidChange calls WebProcessProxy&apos;s
updateSharedPreferencesForWebProcess, which in turn determines if the shared preferences for this
WebProcessProxy has changed or not. If it did, it returns the updated SharedPreferencesForWebProcess,
which is then relayed to the GPU process via GPUProcess::sharedPreferencesForWebProcessDidChange.
Finally, WebPage::preferencesDidChange in WebProcess synchronously waits for
SharedPreferencesForWebProcess to be propagated to the GPU process. For this purpose,
each SharedPreferencesForWebProcessDidChange contains a uint64_t version, which gets incremented
whenever the preferences is updated.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::m_sharedPreferencesForWebProcess):
(WebKit::GPUConnectionToWebProcess::allowsExitUnderMemoryPressure const):
(WebKit::GPUConnectionToWebProcess::createGPU):
(WebKit::m_preferences): Deleted.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::sharedPreferencesForWebProcess const):
(WebKit::GPUConnectionToWebProcess::updateSharedPreferencesForWebProcess):
(WebKit::GPUConnectionToWebProcess::isWebXREnabled const):
(WebKit::GPUConnectionToWebProcess::isDynamicContentScalingEnabled const):
(WebKit::GPUConnectionToWebProcess::isWebGPUEnabled const): Deleted.
(WebKit::GPUConnectionToWebProcess::isWebGLEnabled const): Deleted.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::sharedPreferencesForWebProcessDidChange):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/Scripts/PreferencesTemplates/SharedPreferencesForWebProcess.cpp.erb: Added.
* Source/WebKit/Scripts/PreferencesTemplates/SharedPreferencesForWebProcess.h.erb: Added.
* Source/WebKit/Scripts/PreferencesTemplates/SharedPreferencesForWebProcess.serialization.in.erb: Added.
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
* Source/WebKit/Shared/GPUProcessConnectionParameters.h:
* Source/WebKit/Shared/GPUProcessConnectionParameters.serialization.in:
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.serialization.in: Removed.
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::preferencesForGPUProcess const): Deleted.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::sharedPreferencesForWebProcessDidChange):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::hasSameGPUAndNetworkProcessPreferencesAs const):
(WebKit::WebPageProxy::preferencesDidChange):
(WebKit::WebPageProxy::preferencesForGPUProcess const): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::~WebProcessProxy):
(WebKit::WebProcessProxy::initializePreferencesForGPUAndNetworkProcesses):
(WebKit::WebProcessProxy::hasSameGPUAndNetworkProcessPreferencesAs const):
(WebKit::WebProcessProxy::createGPUProcessConnection):
(WebKit::WebProcessProxy::updateSharedPreferencesForWebProcess):
(WebKit::WebProcessProxy::didSyncSharedPreferencesForWebProcess):
(WebKit::WebProcessProxy::waitForSharedPreferencesForWebProcessToSync):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::sharedPreferencesForWebProcess const):
(WebKit::WebProcessProxy::preferencesForGPUProcess const): Deleted.
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::preferencesDidChange):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::preferencesDidChangeDuringDOMPrintOperation):
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/281360@main">https://commits.webkit.org/281360@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39f011113ef58384b2018041e929925dd28ea54f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59646 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38991 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12171 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63562 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10170 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46644 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10324 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48391 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7123 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61676 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36392 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51635 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29229 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/59173 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33096 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8882 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9093 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/52740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55046 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9160 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65293 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/58891 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3574 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/9051 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55731 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3585 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51623 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55861 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2966 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/80648 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8910 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34805 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13991 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35888 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36974 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35633 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
```diff
@@ -137,6 +137,7 @@ class GPUConnectionToWebProcess
 
+    const WebFeatures& webFeatures() const { return m_preferences.features; }
     bool isWebGPUEnabled() const { return m_preferences.isWebGPUEnabled; }
     bool isWebGLEnabled() const { return m_preferences.isWebGLEnabled; }
```
Can we remove this?

rniwa replied:
Sure, removed.

achristensen07 commented on Source/WebKit/Scripts/PreferencesTemplates/WebFeatures.h.erb:
```diff
@@ -0,0 +1,65 @@
+class Encoder;
+class Decoder;
+
+template<> struct ArgumentCoder<WebKit::WebFeatures> {
```
Does TestWebKitAPI.IPCTestingAPI.SerializedTypeInfo pass after this in debug builds?  I think we probably need to generate a serialization.in file instead.

rniwa replied:
Changed it to generate serialization.in file.

achristensen07 commented on Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.serialization.in:
```diff
@@ -22,7 +22,9 @@
 
+header: "WebFeatures.h"
 struct WebKit::GPUProcessPreferencesForWebProcess {
+    WebKit::WebFeatures features;
```
I think this is missing code to go to each WebPageProxy and get the enabled features on each, and also code to update it if it is updated.  See how we get the value for isWebGLEnabled()

rniwa replied:
? `PageConfiguration::preferencesForGPUProcess()` does that conversion. We're also not updating these preferences boolean anywhere as far as I can tell. It might be a nice enhancement in the future but I don't think it's needed for the initial version.

achristensen07 commented on Source/WebKit/DerivedSources-input.xcfilelist:
```diff
@@ -178,6 +178,11 @@ $(PROJECT_DIR)/Resources/cocoa/NotificationAllowList/EmbeddedForwardedNotificati
 $(PROJECT_DIR)/Resources/cocoa/NotificationAllowList/MacForwardedNotifications.def
 $(PROJECT_DIR)/Resources/cocoa/NotificationAllowList/NonForwardedNotifications.def
+$(PROJECT_DIR)/Scripts/PreferencesTemplates/WebFeatires.cpp.erb
+$(PROJECT_DIR)/Scripts/PreferencesTemplates/WebFeatires.h.erb
```
This file doesn't exist.

cdumez replied:
Typo in the name: WebFeatires.h.erb

achristensen07 commented on Source/WebKit/DerivedSources-input.xcfilelist:
```diff
@@ -691,3 +696,4 @@ $(WEBCORE_PRIVATE_HEADERS_DIR)/ProtectionSpaceBase.serialization.in
 $(WEBCORE_PRIVATE_HEADERS_DIR)/ScrollTypes.serialization.in
 $(WEBCORE_PRIVATE_HEADERS_DIR)/WebGPU.serialization.in
 $(WEBCORE_PRIVATE_HEADERS_DIR)/generate-bindings.pl
+WebFeatures.serialization.in
```
This isn't input, it's output.

achristensen07 commented on Source/WebKit/DerivedSources.make:
```diff
@@ -812,6 +815,7 @@ SERIALIZATION_DESCRIPTION_FILES = \
 	WebProcess/UserContent/InjectUserScriptImmediately.serialization.in \
 	WebProcess/WebCoreSupport/WebSpeechSynthesisVoice.serialization.in \
 	WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemoteProperties.serialization.in \
+    WebFeatures.serialization.in \
```
Indentation.

cdumez commented on Source/WebKit/DerivedSources.make:
```diff
@@ -812,6 +815,7 @@ SERIALIZATION_DESCRIPTION_FILES = \
 	WebProcess/UserContent/InjectUserScriptImmediately.serialization.in \
 	WebProcess/WebCoreSupport/WebSpeechSynthesisVoice.serialization.in \
 	WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemoteProperties.serialization.in \
+    IPCFiltering.serialization.in \
```
nit: indentation may be wrong here.

rniwa replied:
Fixed

cdumez commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
```diff
@@ -448,6 +454,7 @@ class GPUConnectionToWebProcess
     // GPU preferences don't change for a given WebProcess. Pages that use different GPUProcessPreferences
     // cannot be in the same WebProcess.
     const GPUProcessPreferencesForWebProcess m_preferences;
```
Is this still needed?

cdumez replied:
I thought I had added `GPUProcessPreferencesForWebProcess` purely for the purpose of IPC filtering.

rniwa replied:
There are still a few preferences there:

 - bool isDOMRenderingEnabled { false };
 - bool isDynamicContentScalingEnabled { false };
 - bool allowTestOnlyIPC { false };
 - bool isLockdownSafeFontParserEnabled { false };

cdumez replied:
Seems annoying to have 2 structs for similar things. We essentially want to share a setting to the GPUProcess or the NetworkProcess. Maybe `filtersIPC` should be renamed to `shareWith: NetworkProcess|GPUProcess` and then replace `GPUProcessPreferencesForWebProcess ` (and the network equivalent) entirely. Or something like `crossProcess: true` if we feel we don't need to distinguish GPUProcess and NetworkProcess and just share these particular settings with both of them.

rniwa replied:
We can eventually move to always using this new mechanism but I don't want to do that in this PR because it's already pretty big as is. And I think we want these settings to be shared across processes. Differentiating between processes will induce an unnecessary complexity.

cdumez replied:
> Differentiating between processes will induce an unnecessary complexity.
I agree and one of my 2 proposals addressed this.

I'm personally not sure that landing the patch as is would be great. It adds complexity/ambiguity since it adds one more way of doing something, rather than replacing the existing thing. Also, the name of the property in the yaml file (and the name of the generated files) wouldn't make sense if we extended this to properties that are shared with other processes but are NOT necessarily used for IPC filtering.

rniwa replied:
What name do you suggest then? I really don't wanna waste anymore time renaming things. It took me a whole day to do this rename.

cdumez replied:
Since this is a replacement for `GPUProcessPreferencesForWebProcess` and `NetworkProcessProcessPreferencesForWebProcess`, maybe `SharedPreferencesForWebProcess`?

cdumez commented on Source/WebKit/Shared/GPUProcessConnectionParameters.h:
```diff
@@ -41,6 +42,7 @@ namespace WebKit {
 struct GPUProcessConnectionParameters {
     WebCore::ProcessIdentity webProcessIdentity;
     GPUProcessPreferencesForWebProcess preferences;
```
Still needed?

rniwa replied:
Yes.

cdumez reviewed (commented):
Didn't have time to do a full review yet (meetings)

cdumez commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
```diff
@@ -490,7 +490,7 @@ bool GPUConnectionToWebProcess::allowsExitUnderMemoryPressure() const
         return false;
 
-    if (m_preferences.isDOMRenderingEnabled)
+    if (sharedPreferencesForWebProcess().UseGPUProcessForDOMRenderingEnabled)
```
Would be nice if the data members in SharedPreferencesForWebProcess should start with lowercase.

rniwa replied:
Sure, we can change that.

cdumez commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
```diff
@@ -293,7 +293,7 @@ GPUConnectionToWebProcess::GPUConnectionToWebProcess(GPUProcess& gpuProcess, Web
     , m_routingArbitrator(LocalAudioSessionRoutingArbitrator::create(*this))
 #endif
-    , m_preferences(parameters.preferences)
+    , m_sharedPreferencesForWebProcess(parameters.sharedPreferencesForWebProcess)
```
We could WTFMove() here. Not all preferences are booleans so avoiding the copy may be helpful in the future.

rniwa replied:
Ok

cdumez commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
```diff
@@ -137,12 +137,17 @@ class GPUConnectionToWebProcess
-    bool isWebGLEnabled() const { return m_preferences.isWebGLEnabled; }
-    bool isWebXREnabled() const { return m_preferences.isWebXREnabled; }
+    const SharedPreferencesForWebProcess& sharedPreferencesForWebProcess() const { return m_sharedPreferencesForWebProcess; }
+    void updateSharedPreferencesForWebProcess(const SharedPreferencesForWebProcess& sharedPreferencesForWebProcess) { m_sharedPreferencesForWebProcess = sharedPreferencesForWebProcess; }
```
Can we take a r-value reference and WTFMove()?

rniwa replied:
Sure, we can do that.

cdumez commented on Source/WebKit/GPUProcess/GPUProcess.cpp:
```diff
@@ -149,6 +148,14 @@ void GPUProcess::createGPUConnectionToWebProcess(WebCore::ProcessIdentifier iden
     m_webProcessConnections.add(identifier, WTFMove(newConnection));
 }
 
+void GPUProcess::sharedPreferencesForWebProcessDidChange(WebCore::ProcessIdentifier identifier, const SharedPreferencesForWebProcess& sharedPreferencesForWebProcess, CompletionHandler<void()>&& completionHandler)
```
`SharedPreferencesForWebProcess&&`

rniwa replied:
Ok

cdumez commented on Source/WebKit/GPUProcess/GPUProcess.cpp:
```diff
@@ -149,6 +148,14 @@ void GPUProcess::createGPUConnectionToWebProcess(WebCore::ProcessIdentifier iden
+{
+    auto it = m_webProcessConnections.find(identifier);
+    if (it != m_webProcessConnections.end())
+        it->value->updateSharedPreferencesForWebProcess(sharedPreferencesForWebProcess);
```
`WTFMove(sharedPreferencesForWebProcess)`

cdumez replied:
I think this whole function would look nicer and safer like so:
```
if (RefPtr connection = m_webProcessConnections.get(identifier))
    connection->updateSharedPreferencesForWebProcess(WTFMove(sharedPreferencesForWebProcess));
```

rniwa replied:
Sure, we can do that.

cdumez commented on Source/WebKit/GPUProcess/GPUProcess.h:
```diff
@@ -156,6 +156,7 @@ class GPUProcess : public AuxiliaryProcess, public ThreadSafeRefCounted<GPUProce
     void platformInitializeGPUProcess(GPUProcessCreationParameters&);
     void updateGPUProcessPreferences(GPUProcessPreferences&&);
     void createGPUConnectionToWebProcess(WebCore::ProcessIdentifier, PAL::SessionID, IPC::Connection::Handle&&, GPUProcessConnectionParameters&&, CompletionHandler<void()>&&);
+    void sharedPreferencesForWebProcessDidChange(WebCore::ProcessIdentifier, const SharedPreferencesForWebProcess&, CompletionHandler<void()>&&);
```
`SharedPreferencesForWebProcess&&`

rniwa replied:
Fixed.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -1178,14 +1178,12 @@ void WebPageProxy::handleSynchronousMessage(IPC::Connection& connection, const S
-    if (preferencesForGPUProcess() != configuration.preferencesForGPUProcess())
-        return false;
-#endif
     if (preferencesForNetworkProcess() != configuration.preferencesForNetworkProcess())
```
Are you planning to get rid of `preferencesForNetworkProcess` in a follow-up?

rniwa replied:
Yes.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5879,10 +5877,23 @@ void WebPageProxy::preferencesDidChange()
+
         if (m_isPerformingDOMPrintOperation)
-            webProcess.send(Messages::WebPage::PreferencesDidChangeDuringDOMPrintOperation(preferencesStore()), pageID,  IPC::SendOption::DispatchMessageEvenWhenWaitingForUnboundedSyncReply);
+            webProcess.send(Messages::WebPage::PreferencesDidChangeDuringDOMPrintOperation(preferencesStore(), sharedPreferencesVersion), pageID,  { IPC::SendOption::IgnoreFullySynchronousMode, IPC::SendOption::DispatchMessageEvenWhenWaitingForUnboundedSyncReply });
```
Would be good to explain these IgnoreFullySynchronousMode / IgnoreFullySynchronousMode changes in the commit log. They're not-obvious to me at least.

rniwa replied:
Oh, this change is no longer needed now that I'm not including setPrintBackgrounds as one of preferences. Will revert.

cdumez replied:
Great, that's what I was hoping.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -783,9 +786,12 @@ class WebProcessProxy : public AuxiliaryProcessProxy {
+    mutable std::optional<SharedPreferencesForWebProcess> m_sharedPreferencesForWebProcess;
+    uint64_t m_syncedSharedPreferencesVersion { 0 };
+    uint64_t m_awaitedSharedPreferencesVersion { 0 };
+    std::optional<CompletionHandler<void(bool success)>> m_sharedPreferencesForWebProcessCompletionHandler;
```
We should not be using `std::optional<CompletionHandler<void(bool success)>>`. `CompletionHandler<void(bool success)>` is enough since CompletionHandler has a null state. It's the same thing as `std::optional<String>`, it wouldn't make sense since String has a null state.

rniwa replied:
Fixed

cdumez reviewed (approved):
LGTM with changes.

rniwa committed:
Add a mechanism to filter IPC messages by feature flags
https://bugs.webkit.org/show_bug.cgi?id=276415

Reviewed by Chris Dumez.

This PR adds a filtering mechanism for IPC messages based on which features are enabled at runtime.

To do this, this PR adds the code to generate SharedPreferencesForWebProcess.h/cpp/serialization.in,
which consists of a simple struct with a bunch of booleans indicating which preference is enabled
in a given WebContent process. UnifiedWebPreferences.yaml is now annotated with
"sharedPreferenceForWebProcess" flag per preference, and messages.in files now support "EnabledBy=X"
attribute which filters IPC messages by the preferences.

The set of preferences for a given WebContent process is synchronized at the time of constructing
GPUConnectionToWebProcess as a part of GPUProcessConnectionParameters, and updated via
GPUProcessProxy::filteringDidChangeForWebProcess. Because multiple WebPageProxy can share a single
WebContent process, the set is the union of all preferences enabled across those pages.

When a preference is updated, WebPageProxy::preferencesDidChange calls WebProcessProxy's
updateSharedPreferencesForWebProcess, which in turn determines if the shared preferences for this
WebProcessProxy has changed or not. If it did, it returns the updated SharedPreferencesForWebProcess,
which is then relayed to the GPU process via GPUProcess::sharedPreferencesForWebProcessDidChange.
Finally, WebPage::preferencesDidChange in WebProcess synchronously waits for
SharedPreferencesForWebProcess to be propagated to the GPU process. For this purpose,
each SharedPreferencesForWebProcessDidChange contains a uint64_t version, which gets incremented
whenever the preferences is updated.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::m_sharedPreferencesForWebProcess):
(WebKit::GPUConnectionToWebProcess::allowsExitUnderMemoryPressure const):
(WebKit::GPUConnectionToWebProcess::createGPU):
(WebKit::m_preferences): Deleted.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::sharedPreferencesForWebProcess const):
(WebKit::GPUConnectionToWebProcess::updateSharedPreferencesForWebProcess):
(WebKit::GPUConnectionToWebProcess::isWebXREnabled const):
(WebKit::GPUConnectionToWebProcess::isDynamicContentScalingEnabled const):
(WebKit::GPUConnectionToWebProcess::isWebGPUEnabled const): Deleted.
(WebKit::GPUConnectionToWebProcess::isWebGLEnabled const): Deleted.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::sharedPreferencesForWebProcessDidChange):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/Scripts/PreferencesTemplates/SharedPreferencesForWebProcess.cpp.erb: Added.
* Source/WebKit/Scripts/PreferencesTemplates/SharedPreferencesForWebProcess.h.erb: Added.
* Source/WebKit/Scripts/PreferencesTemplates/SharedPreferencesForWebProcess.serialization.in.erb: Added.
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
* Source/WebKit/Shared/GPUProcessConnectionParameters.h:
* Source/WebKit/Shared/GPUProcessConnectionParameters.serialization.in:
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.serialization.in: Removed.
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::preferencesForGPUProcess const): Deleted.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::sharedPreferencesForWebProcessDidChange):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::hasSameGPUAndNetworkProcessPreferencesAs const):
(WebKit::WebPageProxy::preferencesDidChange):
(WebKit::WebPageProxy::preferencesForGPUProcess const): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::~WebProcessProxy):
(WebKit::WebProcessProxy::initializePreferencesForGPUAndNetworkProcesses):
(WebKit::WebProcessProxy::hasSameGPUAndNetworkProcessPreferencesAs const):
(WebKit::WebProcessProxy::createGPUProcessConnection):
(WebKit::WebProcessProxy::updateSharedPreferencesForWebProcess):
(WebKit::WebProcessProxy::didSyncSharedPreferencesForWebProcess):
(WebKit::WebProcessProxy::waitForSharedPreferencesForWebProcessToSync):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::sharedPreferencesForWebProcess const):
(WebKit::WebProcessProxy::preferencesForGPUProcess const): Deleted.
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::preferencesDidChange):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::preferencesDidChangeDuringDOMPrintOperation):
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/281360@main
